File: tests/fsharp/core/csfromfs/indexers.fs
#indent "off"

module Indexers

let failures = ref false
let report_failure () = 
  stderr.WriteLine "NO"; failures := true


let testIndexers = 
  begin 
    let c = new CSharpIndexers.C() in 
    if c.Item(3) <> 203 then report_failure();
    if (upcast c : CSharpIndexers.I).Item(3) <> 103 then report_failure(); 
    let d = new CSharpIndexers.D() in 
    if d.Item(3) <> 303 then report_failure();
    if (upcast d : CSharpIndexers.C).Item(3) <> 303 then report_failure();
  end;
  begin 
    let d = new CSharpIndexers.OverloadedIndexer() in 
    if d.Item(3) <> 203 then report_failure();
    if d.Item("2") <> 201 then report_failure();
    if (upcast d : CSharpIndexers.I2).Item(3) <> 103 then report_failure();
    if (upcast d : CSharpIndexers.I2).Item("4") <> 101 then report_failure();
   
  end


let _ = 
  if !failures then (stdout.WriteLine "Test Failed"; exit 1) 
  else (stdout.WriteLine "Test Passed"; 
        System.IO.File.WriteAllText("test.ok","ok"); 
        exit 0)

File: tests/fsharp/core/csfromfs/indexers.cs

using System; using System.Diagnostics;


//****************************************************************************
// Indexers
//****************************************************************************

namespace CSharpIndexers
{
	public interface I
	{
		int this [int i] {
			get;
			set;
		}
	}

	public interface I2
	{
		int this [int i] {
			get;
			set;
		}
		int this [string i] {
			get;
			set;
		}
	}

	public class C : I
	{

		int I.this [int i] {
			get { return 100 + i; }
			set { return; }
		}

		
		public virtual int this [int i] {
			get { return 200 + i; } set { return; }
		}
	}

	public class D : C
	{
		public override int this [int i] {
                        get { return 300 + i; }
			set { return; }
		}
	}
	public class OverloadedIndexer : I2
	{

		int I2.this [int i] {
			get { return 100 + i; }
			set { return; }
		}

		int I2.this [string i] {
			get { return 100 + i.Length; }
			set { return; }
		}

		
		public virtual int this [int i] {
			get { return 200 + i; } set { return; }
		}
		public virtual int this [string i] {
			get { return 200 + i.Length; } set { return; }
		}
	}

}

